homeworks/hw12/src/bank.rs

```diff
@@ -0,0 +1,869 @@
+    ///
+    /// ```
+    fn replay_history(source_bank: Bank) -> Bank {
+        let mut target_bank = Bank::new();
+
+        let history_result = source_bank.get_history();
+        if let Ok(history) = history_result {
+            for operation in history {
+                match operation.operation_type {
+                    OperationType::CreateAccount => target_bank
+                        .create_account(operation.source_account.as_str())
+                        .unwrap(),
+                    OperationType::Deposit => target_bank
+                        .deposit(operation.source_account.as_str(), operation.amount)
+                        .unwrap(),
+                    OperationType::Withdraw => target_bank
+                        .withdraw(operation.source_account.as_str(), operation.amount)
+                        .unwrap(),
+                    OperationType::Transfer => target_bank
+                        .transfer(
+                            operation.source_account.as_str(),
+                            operation.target_account.unwrap().as_str(),
+                            operation.amount,
+                        )
+                        .unwrap(),
+                };
+            }
+        }
+        target_bank
+    }
+}
+
+pub trait BankTrait {
+    /// Creates a new account with the specified name and adds it to the bank.
+    ///
+    /// # Arguments
+    ///
+    /// * `account` - The code of the account to create.
+    ///
+    /// # Errors
+    /// AccountDuplicationError
+    ///
+    /// Returns an error if an account with the same name already exists in the bank.
+    ///
+    /// ```
+    fn create_account(&mut self, account: &str) -> Result<TransactionId, BankError>;
+    /// Deposits the specified amount into the account.
+    ///
+    /// # Arguments
+    ///
+    /// * `account` - The account to deposit into.
+    /// * `amount` - The amount to deposit into the account.
+    ///
+    /// #Returns
+    ///  TransactionId for the new account
+    ///  BankError if the account does not exist or the amount is negative
+    /// # Errors
+    /// AmountNegativeError
+    /// AccountNotFoundError
+    ///
+    /// ```
+    fn deposit(&mut self, account: &str, amount: Money) -> Result<TransactionId, BankError>;
+    /// Withdraws the specified amount from the account.
+    ///
+    /// # Arguments
+    ///
+    /// * `account` - The account to withdraw from.
+    /// * `amount` - The amount to withdraw from the account.
+    ///
+    /// # Returns
+    /// TransactionId for processing the transaction
+    /// BankError if the account does not exist or the amount is negative or insufficient funds
+    /// # Errors
+    /// AmountNegativeError
+    /// AccountNotFoundError
+    /// InsufficientFundsError
+    ///
+    /// ```
+
+    fn withdraw(&mut self, account: &str, amount: Money) -> Result<TransactionId, BankError>;
+    /// Transfers the specified amount from one account to another.
```

Comment:
```suggestion
    /// ```
    fn withdraw(&mut self, account: &str, amount: Money) -> Result<TransactionId, BankError>;

    /// Transfers the specified amount from one account to another.
```